wp/lib/bap_wp/src/compare.ml

```diff
@@ -113,62 +113,65 @@ let compare_blocks
+    ~original:(sub1, env1 : Sub.t * Env.t)
+    ~modified:(sub2, env2 : Sub.t * Env.t)
+    ~rename_set:(vars : Var.Set.t)
+    (comparators : comparator list)
+  : Constr.t * Env.t * Env.t =
+  let ctx = Env.get_context env1 in
+  let constrs, env1, env2 =
+    List.fold comparators ~init:([], env1, env2)
+      ~f:(fun (comps, env1, env2) comp ->
+          let c, env1, env2 =
+            comp ~original:(sub1, env1) ~modified:(sub2, env2) ~rename_set:vars
+          in
+          (* We are not adding the trivial constraints to prevent blow up in
+             the final precondition. *)
+          if Bool.is_true (Constr.eval c ctx) then
+            comps, env1, env2
+          else
+            c :: comps, env1, env2)
+  in
+  Constr.mk_clause [] constrs, env1, env2
+
 (* A generic function for generating a predicate that compares two subroutines.
    Takes as input a postcondition generator, and a hypothesis generator.
    this could be made even more generic, say by abstracting over the type of terms
    being compared. *)
 let compare_subs
-    ~postcond:(postcond : comparator)
-    ~hyps:(hyps : comparator)
+    ~postconds:(postconds : comparator list)
+    ~hyps:(hyps : comparator list)
     ~original:(sub1, env1 : Sub.t * Env.t)
     ~modified:(sub2, env2 : Sub.t * Env.t)
-    ~smtlib_post:(smtlib_post : string)
-    ~smtlib_hyp:(smtlib_hyp : string)
   : Constr.t * Env.t * Env.t =
   let env2 = Env.set_freshen env2 true in
   let vars = Var.Set.union
       (Pre.get_vars env1 sub1) (Pre.get_vars env2 sub2) in
   let post, env1, env2 =
-    let post, env1, env2 =
-      postcond ~original:(sub1, env1) ~modified:(sub2, env2) ~rename_set:vars in
-    let smtlib_post = mk_smtlib2_compare env1 env2 smtlib_post in
-    Constr.mk_clause [] [post; smtlib_post], env1, env2
+    fold_comparators postconds ~original:(sub1, env1) ~modified:(sub2, env2)
+      ~rename_set:vars
   in
   info "\nPostcondition:\n%s\n%!" (Constr.to_string post);
   let hyps, env1, env2 =
-    let hyps, env1, env2 =
-      hyps ~original:(sub1, env1) ~modified:(sub2, env2) ~rename_set:vars in
-    let smtlib_hyp = mk_smtlib2_compare env1 env2 smtlib_hyp in
-    let sp_range = Pre.set_sp_range env1 in
-    let init_mem, env1, env2 = init_vars env1 env2 vars in
-    Constr.mk_clause [] (hyps :: smtlib_hyp :: sp_range :: init_mem), env1, env2
+    fold_comparators hyps ~original:(sub1, env1) ~modified:(sub2, env2)
+      ~rename_set:vars
   in
   info "\nHypotheses:\n%s\n%!" (Constr.to_string hyps);
   let pre_mod, env2 = Pre.visit_sub env2 post sub2 in
   let pre_combined, env1 = Pre.visit_sub env1 pre_mod sub1 in
   let goal = Constr.mk_clause [hyps] [pre_combined] in
   goal, env1, env2
 
-let compare_subs_empty
-    ~original:(original : Sub.t * Env.t)
-    ~modified:(modified : Sub.t * Env.t)
-    ~smtlib_post:(smtlib_post : string)
-    ~smtlib_hyp:(smtlib_hyp : string)
-  : Constr.t * Env.t * Env.t =
+let compare_subs_empty () : comparator * comparator =
```

Comment:
Why does this take unit as input?

Reply:
No particular reason. I can remove it.

Reply:
Yea, unit is only useful to delay computation, if something non-trivial is happening in the body, or some side-effect occurs. Neither is the case here.